src/Microsoft.Extensions.Configuration.AzureKeyVault/AzureKeyVaultConfigurationProvider.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Threading.Tasks;
+using Microsoft.Azure.KeyVault;
+
+namespace Microsoft.Extensions.Configuration.AzureKeyVault
+{
+    /// <summary>
+    /// An AzureKeyVault based <see cref="ConfigurationProvider"/>.
+    /// </summary>
+    public class AzureKeyVaultConfigurationProvider : ConfigurationProvider
+    {
+        private readonly IKeyVaultClient _client;
+        private readonly string _vault;
+        private readonly Func<SecretItem, bool> _filter;
+
+        /// <summary>
+        /// Creates a new instance of <see cref="AzureKeyVaultConfigurationProvider"/>.
+        /// </summary>
+        /// <param name="client">The <see cref="KeyVaultClient"/> to use for retrieving values.</param>
+        /// <param name="vault">Azure KeyVault uri.</param>
+        /// <param name="filter">The predicate to filter secret entries before loading value, <code>null</code> to load all.</param>
+        public AzureKeyVaultConfigurationProvider(IKeyVaultClient client, string vault, Func<SecretItem, bool> filter)
+        {
+            if (client == null)
+            {
+                throw new ArgumentNullException(nameof(client));
+            }
+            if (vault == null)
+            {
+                throw new ArgumentNullException(nameof(vault));
+            }
+            _client = client;
+            _vault = vault;
+            _filter = filter;
+        }
+
+        /// <inheritdoc />
+        public override void Load()
+        {
+            LoadAsync().GetAwaiter().GetResult();
+        }
+
+        private async Task LoadAsync()
+        {
+            var sec = await _client.GetSecretsAsync(_vault);
+            do
+            {
+                foreach (var secretItem in sec.Value)
+                {
+                    if (_filter?.Invoke(secretItem) == false)
+                    {
+                        continue;
+                    }
+
+                    var value = await _client.GetSecretAsync(secretItem.Id);
+                    Data.Add(value.SecretIdentifier.Name, value.Value);
```

Comment:
This will throw if the key already exists (e.g. Load() is invoked twice) - is it OK or the value for the key should be overwritten?